adding examples in schema transforms section of programming guide for python  (changes for issue #21022) (#23224)
apache/beam: website/www/site/content/en/documentation/programming-guide.md

 shipping address one would write
{{< /paragraph >}}

{{< highlight java >}}
purchases.apply(Select.fieldNames("shippingAddress.postCode"));
{{< /highlight >}}
	
<!-- {{< highlight py >}} 
input_pc = ... # {"user_id": ..., "shipping_address": "post_code": ..., "bank": ..., "purchase_amount": ...}
output_pc = input_pc | beam.Select(post_code=lambda item: str(item["shipping_address.post_code"]))                                               
{{< /highlight >}} -->
{{< paragraph class="language-java" >}}
##### **Wildcards**
{{< /paragraph >}}

{{< paragraph class="language-java" >}}
The * operator can be specified at any nesting level to represent all fields at that level. For example, to select all
shipping-address fields one would write
{{< /paragraph >}}

{{< highlight java >}}
purchases.apply(Select.fieldNames("shippingAddress.*"));
{{< /highlight >}}

<!-- {{< highlight py >}} 
input_pc = ... # {"user_id": ..., "shipping_address": "post_code": ..., "bank": ..., "purchase_amount": ...}
output_pc = input_pc | beam.Select(shipping_address=lambda item: str(item["shipping_address.*"]))                                               

> {{< /highlight >}} -->

Reviewer: This the section we want to add in later, right? It may be better to have TODO along with the issue number (https://github.com/apache/beam/issues/23275) instead of this commented-out code.

Author: I agree

Reviewer 2: ```suggestion
<!-- 
{{< highlight py >}} 
#TODO(https://github.com/apache/beam/issues/23275): Add support for projecting nested fields
input_pc = ... # {"user_id": ..., "shipping_address": "post_code": ..., "bank": ..., "purchase_amount": ...}
output_pc = input_pc | beam.Select("shipping_address.*"))                                               
{{< /highlight >}} -->
```

purchases.apply(Select.fieldNames("userId"));
{{< /highlight >}}

> {{< highlight python >}}

Reviewer 2: I think this is the issue. Unfortunately our infra is considering "python" to be a separate language from "py"

```suggestion
{{< highlight py >}}
```

Author: ohh thanks

input_pc = ... # {"user_id": ...,"bank": ..., "purchase_amount": ...}

> output_pc = input_pc | beam.Select(user_id=lambda item: str(item["user_id"]))

Reviewer 2: Let's recommend the string argument approach here:
```suggestion
output_pc = input_pc | beam.Select("user_id")
```

{{< /highlight >}} -->
##### **Wildcards**

{{< paragraph class="language-py" >}}
Support for wildcards hasn't been developed for python SDK yet
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for wildcards hasn't been developed for GO SDK yet
{{< /paragraph >}}

{{< paragraph class="language-java" >}}
The * operator can be specified at any nesting level to represent all fields at that level. For example, to select all
shipping-address fields one would write
{{< /paragraph >}}

{{< highlight java >}}
purchases.apply(Select.fieldNames("shippingAddress.*"));
{{< /highlight >}}

<!-- {{< highlight py >}} 
input_pc = ... # {"user_id": ..., "shipping_address": "post_code": ..., "bank": ..., "purchase_amount": ...}
output_pc = input_pc | beam.Select(shipping_address=lambda item: str(item["shipping_address.*"]))                                               
{{< /highlight >}} -->
##### **Arrays**

{{< paragraph class="language-java" >}}
An array field, where the array element type is a row, can also have subfields of the element type addressed. When
selected, the result is an array of the selected subfield type. For example
{{< /paragraph >}}

> {{< paragraph class="language-py" >}}
> Support for Nested fields hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for Nested fields hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for Array fields hasn't been developed for python SDK yet
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for Array fields hasn't been developed for GO SDK yet
{{< /paragraph >}}
```

purchases.apply(Select.fieldNames("userId"));
{{< /highlight >}}

{{< highlight py >}}
input_pc = ... # {"user_id": ...,"bank": ..., "purchase_amount": ...}
output_pc = input_pc | beam.Select("user_id")
{{< /highlight >}}

##### **Nested fields**

> {{< paragraph class="language-py" >}}
> Support for Nested fields hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for Nested fields hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for Nested fields hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for Nested fields hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```

{{< paragraph class="language-java" >}}
Individual nested fields can be specified using the dot operator. For example, to select just the postal code from the
 shipping address one would write
{{< /paragraph >}}

{{< highlight java >}}
purchases.apply(Select.fieldNames("shippingAddress.postCode"));
{{< /highlight >}}
	
<!-- {{< highlight py >}} 
input_pc = ... # {"user_id": ..., "shipping_address": "post_code": ..., "bank": ..., "purchase_amount": ...}
output_pc = input_pc | beam.Select(post_code=lambda item: str(item["shipping_address.post_code"]))                                               
{{< /highlight >}} -->
##### **Wildcards**

> {{< paragraph class="language-py" >}}
> Support for wildcards hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for wildcards hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for wildcards hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for wildcards hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```

{{< paragraph class="language-java" >}}
The * operator can be specified at any nesting level to represent all fields at that level. For example, to select all
shipping-address fields one would write
{{< /paragraph >}}

{{< highlight java >}}
purchases.apply(Select.fieldNames("shippingAddress.*"));
{{< /highlight >}}

<!-- 
{{< highlight py >}} 
#TODO(https://github.com/apache/beam/issues/23275): Add support for projecting nested fields
input_pc = ... # {"user_id": ..., "shipping_address": "post_code": ..., "bank": ..., "purchase_amount": ...}
output_pc = input_pc | beam.Select("shipping_address.*"))                                               
{{< /highlight >}} -->
##### **Arrays**

{{< paragraph class="language-java" >}}
An array field, where the array element type is a row, can also have subfields of the element type addressed. When
selected, the result is an array of the selected subfield type. For example
{{< /paragraph >}}

> {{< paragraph class="language-py" >}}
> Support for Array fields hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for Array fields hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for Array fields hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for Array fields hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```

purchases.apply(Select.fieldNames("userId"));
{{< /highlight >}}

{{< highlight py >}}

> input_pc = ... # {"user_id":...,"bank": ..., "purchase_amount": ...}

Reviewer: ```suggestion
input_pc = ... # {"user_id": ...,"bank": ..., "purchase_amount": ...}
```

Author: I was actually thinking about the same, if we could represent it in a kind off pretty json format it would provide better understandability of the input

purchases.apply(Select.fieldNames("userId"));
{{< /highlight >}}

{{< highlight py >}}

> input_pc = ... # {"userId":...,"bank": ..., "purchase_amount": ...}
> output_pc = input_pc | beam.Map(lambda item: beam.Row(ids=str(item["userId"])))                                            

Reviewer: ```suggestion
input_pc = ... # {"user_id":...,"bank": ..., "purchase_amount": ...}
output_pc = input_pc | beam.Select(user_id=lambda item: str(item["user_id"]))                                            
```

Author: Thanks for guiding ,I'll make the appropriate changes

{{< highlight java >}}
purchases.apply(Select.fieldNames("shippingAddress.postCode"));
{{< /highlight >}}
	
{{< highlight py >}} 

> input_pc = ... # {"user_id":...,"shipping_address":"post_code": ..., ...,"bank": ..., "purchase_amount": ...}

Reviewer: ```suggestion
input_pc = ... # {"user_id": ..., "shipping_address":"post_code": ..., ...,"bank": ..., "purchase_amount": ...}
```

{{< highlight java >}}
purchases.apply(Select.fieldNames("shippingAddress.postCode"));
{{< /highlight >}}
	
{{< highlight py >}} 

> input_pc = ... # {"userId":...,"shippingAddress":"postCode": ..., ...,"bank": ..., "purchase_amount": ...}

Reviewer: ```suggestion
input_pc = ... # {"user_id":...,"shipping_address":"postCode": ..., ...,"bank": ..., "purchase_amount": ...}
```

> output_pc = input_pc | beam.Map(lambda item: beam.Row(pcode=str(item["shippingAddress.postCode"])))                                               

Reviewer: Please use `Select`

purchases.apply(Select.fieldNames("shippingAddress.*"));
{{< /highlight >}}

{{< highlight py >}} 

> input_pc = ... # {"user_id":...,"shipping_address":"post_code": ..., ...,"bank": ..., "purchase_amount": ...}

Reviewer: ```suggestion
input_pc = ... # {"user_id": ..., "shipping_address":"post_code": ..., ...,"bank": ..., "purchase_amount": ...}
```

purchasesByType.apply(Select.fieldNames("purchases{}.userId"));
{{< /highlight >}}

> {{< paragraph class="language-py" >}}
> Support for Nested fields hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for Nested fields hasn't been developed for python SDK yet

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for Map fields hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for Map fields hasn't been developed for the Go SDK yet.
```

Reviewer 2: nit: in general can we make these paragraphs "Support for X hasn't been developed for the {Python,Go} SDK yet."

Author: ohh thanks I'll keep that in mind

purchases.apply(Select.fieldNames("userId", "shippingAddress.streetAddress"));
{{< /highlight >}}

> {{< paragraph class="language-py" >}}
> Support for Nested fields hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for Nested fields hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for Nested fields hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for Nested fields hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```

purchases.apply(Select.fieldNames("userId", "shippingAddress.*"));
{{< /highlight >}}

> {{< paragraph class="language-py" >}}
> Support for Wildcards hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for Wildcards hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for Wildcards hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for Wildcards hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```

purchases.apply(Select.fieldNames( "transactions.bank", "transactions.purchaseAmount"));
{{< /highlight >}}

> {{< paragraph class="language-py" >}}
> Support for nested fields hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for nested hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for nested fields hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for nested fields hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```

purchases.apply(Select.flattenedSchema());
{{< /highlight >}}

> {{< paragraph class="language-py" >}}
> Support for nested fields hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for nested fields hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for nested fields hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for nested fields hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```

{{< /paragraph >}}

{{< paragraph class="language-py" >}}
The `GroupBy` transform allows simply grouping data by any number of fields in the input schema, applying aggregations to
those groupings, and storing the result of those aggregations in a new schema field. The output of the `GroupBy` transform
has a schema with one field corresponding to each aggregation performed.
{{< /paragraph >}}

{{< paragraph class="language-java" >}}
The simplest usage of `Group` specifies no aggregations, in which case all inputs matching the provided set of fields
are grouped together into an `ITERABLE` field. For example
{{< /paragraph >}}

{{< paragraph class="language-py" >}}
The simplest usage of `GroupBy` specifies no aggregations, in which case all inputs matching the provided set of fields
are grouped together into an `ITERABLE` field. For example
{{< /paragraph >}}

{{< highlight java >}}
purchases.apply(Group.byFieldNames("userId", "bank"));
{{< /highlight >}}

{{< highlight py >}}
input_pc = ... # {"user_id": ...,"bank": ..., "purchase_amount": ...}
output_pc = input_pc | beam.GroupBy('user_id','bank')
{{< /highlight >}}

> {{< paragraph class="language-go" >}}
> Support for nested fields hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-go" >}}
Support for schema-aware grouping hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```

  </tbody>
</table>
<br/>
{{< /paragraph >}}

{{< paragraph class="language-py" >}}
<table>
  <thead>
    <tr class="header">
      <th><b>Field Name</b></th>
      <th><b>Field Type</b></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>key</td>
      <td>ROW{userId:STRING, bank:STRING}</td>
    </tr>
    <tr>
      <td>values</td>
      <td>ITERABLE[ROW[Purchase]]</td>
    </tr>
  </tbody>
</table>
<br/>

> {{< /paragraph >}}

Reviewer 2: It looks like our paragraph shortcode doesn't work when it has a table in it, both output schemas are shown:
![image](https://user-images.githubusercontent.com/675055/198152209-9dc9785e-7767-4f41-98d5-b9992b91b12d.png)

How about we just update the Java example to use "bank" instead of the nested field? Then we can use the same output schema for both.

Author: okay i'll update the java example to use bank and make appropriate changes for another table below

    .aggregateField("costCents", Top.<Long>largestLongsFn(10), "topPurchases"));
{{< /highlight >}}

{{< highlight py >}}

> input_pc = ... # {"user_id": ...,"item_Id": ..., "cost_cents": ...}
> output_pc = input_pc | beam.GroupBy("user_id")
> 	.aggregate_field("item_id",CountCombineFn,"num_purchases")
> 	.aggregate_field("cost_cents",sum,"total_spendcents")
> 	.aggregate_field("cost_cents",TopCombineFn,"top_purchases")	

Reviewer 2: ```suggestion
input_pc = ... # {"user_id": ..., "item_Id": ..., "cost_cents": ...}
output_pc = input_pc | beam.GroupBy("user_id")
	.aggregate_field("item_id", CountCombineFn, "num_purchases")
	.aggregate_field("cost_cents", sum, "total_spendcents")
	.aggregate_field("cost_cents", TopCombineFn, "top_purchases")	
```

{{< /highlight >}}

> {{< paragraph class="language-go" >}}
> Support for nested fields hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-go" >}}
Support for schema-aware grouping hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```

    .aggregateField("costCents", Top.<Long>largestLongsFn(10), "topPurchases"));
{{< /highlight >}}

{{< highlight python >}}
input_pc = ... # {"user_id": ...,"item_Id": ..., "cost_cents": ...}
output_pc = input_pc | beam.GroupBy("user_id")
	.aggregate_field("item_id",count,"num_purchases")

> 	.aggregate_field("cost_cents",sum,"total_spendcents")

Reviewer 2: nit: `sum` will work here since it's a built-in that we detect, but you need to define `count`. What you want here is the CombineFn used in the [`Count` transform](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.combiners.html#apache_beam.transforms.combiners.Count).

Unfortunately it looks like we don't expose the CombineFn like Java does via `Count.combineFn`. Unless I'm missing something.. do you know @robertwb or @tvalentyn?

Author: So I can't use count directly? (other than defining the count function)
are there any other aggregate functions that can be used which need not be defined? I guess we can use max/min here

Reviewer 3: there is https://github.com/apache/beam/blob/master/sdks/python/apache_beam/transforms/combiners.py#L158 , but I am not sure what exactly is needed to 'expose' this CombineFn - perhaps you could file a FR @TheNeuralBit ?

Reviewer 2: Good point, I filed https://github.com/apache/beam/issues/23511

"natural join" - one in which the same field names are used on both the left-hand and right-hand sides of the join -
and is specified with the `using` keyword:

> {{< paragraph class="language-py" >}}
> Support for joins hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for joins hasn't been developed for GO SDK yet
> {{< /paragraph >}}
> 

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for joins hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for joins hasn't been developed for the Go SDK yet.
{{< /paragraph >}}

```

  </tbody>
</table>
<br/>
{{< /paragraph >}}

{{< paragraph class="language-py" >}}
The result of this aggregation will have the following schema:
<table>
  <thead>
    <tr class="header">
      <th><b>Field Name</b></th>
      <th><b>Field Type</b></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>key</td>
      <td>ROW{userId:STRING}</td>
    </tr>
    <tr>
      <td>value</td>

>       <td>ROW{num_purchases: INT64, total_spendcents: INT64, top_pdghurchases: ARRAY[INT64]}</td>

Reviewer 2: ```suggestion
      <td>ROW{num_purchases: INT64, total_spendcents: INT64, top_purchases: ARRAY[INT64]}</td>
```

  </tbody>
</table>
<br/>
{{< /paragraph >}}

Each resulting row contains one Transaction and one Review that matched the join condition.

If the fields to match in the two schemas have different names, then the on function can be used. For example, if the
Review schema named those fields differently than the Transaction schema, then we could write the following:

> {{< paragraph class="language-py" >}}
> Support for joins hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for joins hasn't been developed for GO SDK yet
> {{< /paragraph >}}
> 

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for joins hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for joins hasn't been developed for the Go SDK yet.
{{< /paragraph >}}

```

"natural join" - one in which the same field names are used on both the left-hand and right-hand sides of the join -

> and is specified with the `using` keyword:

Reviewer 2: Should this paragraph be in a language-java block? Similarly for the other Java-only schema transforms - it seems theres a lot of language that needs to be hidden when Python/Go are selected.

Reviewer 2: If you'd rather not open that can of worms (there are a lot of schema transforms...), we could also just edit the Select/GroupBy-relevant sections in this PR, and leave the other ones as-is for now (and address in a follow-up PR).

Author: I think we should only edit the Select/GroupBy sections in this PR and create another PR, it's getting difficult to keep track of all the sections

Reviewer 2: Agreed

processed in unexpanded format - providing the join key along with Iterables of all records from each input that matched
that key.

> {{< paragraph class="language-py" >}}
> Support for joins hasn't been developed for python SDK yet
> {{< /paragraph >}}
> 
> {{< paragraph class="language-go" >}}
> Support for joins hasn't been developed for GO SDK yet
> {{< /paragraph >}}

Reviewer 2: ```suggestion
{{< paragraph class="language-py" >}}
Support for joins hasn't been developed for the Python SDK yet.
{{< /paragraph >}}

{{< paragraph class="language-go" >}}
Support for joins hasn't been developed for the Go SDK yet.
{{< /paragraph >}}
```
